### PR TITLE
Feature/57 from file parser

### PR DESCRIPTION
**Pull request by dk1844**
From file parser implementation.

Self explanatory tests added (string-based, example file based)

### CODE REVIEW

File: ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  def parseString(str: String): DBFunctionFromSource = {
+    import PgFunctionFileParser._
+
+    val owner = ownerRx.findAllMatchIn(str).toList match {
+      case Nil => throw new IllegalStateException(s"Could not parse owner from $str")
+      case o :: Nil => o.group(1)
+      case _ => throw new IllegalStateException(s"Found more than 2 owners in $str")
+    }
```
Comment (benedeki) [inline ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala:42]: Would put each of these blocks into its own private function.

File: ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    val (databaseName, users) = dbUsersRx.findAllMatchIn(str).toList match {
+      case Nil => throw new IllegalStateException(s"Could not parse database or users from $str")
+      case line :: Nil => {
+        val dbName = line.group(1)
+        val unparsedUsers = line.group(2)
+        val trimmedUsers = unparsedUsers.split(',').map(_.trim).filter(_.nonEmpty)
+        if (trimmedUsers.isEmpty) {
```
[the remark this answers is not in the document]
Comment (dk1844) [reply ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala]: Thanks, fixed.

File: ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      case _ => throw new IllegalStateException(s"Found more than 2 create function mentions in $str")
+    }
+
+    DBFunctionFromSource(fnName, inParamTypes, owner, users, schemaName, databaseName, str)
+  }
+}
+
+object PgFunctionFileParser {
+
+  val ownerRx = """--\s*owner:\s*([_a-zA-Z0-9]+)\s*""".r
```
[the remark this answers is not in the document]
Comment (dk1844) [reply ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala]: Done, thanks for the suggestion.

File: ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+        val paramsUnparsed = paramsStr.split(',').map(_.trim).toSeq
+        val optParamTypes: Seq[Option[String]] = paramsUnparsed.map { param =>
+          param match {
+            case singleParamCapturingRx(inOut, _, paramType) =>
+              if (inOut.toUpperCase == "IN") Some(paramType) else None
+
+            case _ => throw new IllegalStateException(s"Could not parameter from $param")
+          }
+        }
```
Comment (Zejnilovic) [inline ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala:75]:
```suggestion
        val optParamTypes: Seq[Option[String]] = paramsUnparsed.map {
          case singleParamCapturingRx(inOut, _, paramType) =>
            if (inOut.toUpperCase == "IN") Some(paramType) else None

          case param => throw new IllegalStateException(s"Could not parameter from $param")
        }
```

**Review by Zejnilovic — APPROVED** [review]

File: ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala (hunk 1 above)
Comment (dk1844) [reply ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala]: Sure, thanks, done.

File: ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala (hunk 4 above)
Comment (dk1844) [reply ultet/src/main/scala/za/co/absa/ultet/parsers/PgFunctionFileParser.scala]: Thanks, done.

**Pull request merged**